### PR TITLE
Fix handling of short discriminator in Linux BLE scan.

### LINKED ISSUE #21160: Chip-tool wrongly handle discovery via BLE with manual pairing code
#### Problem
The Manual Pairing code contains the information about 4 most significant bits of discriminator.
The current implementation of chip-tool commissioning via BLE wrongly handles the discovery of devices that have set the discriminator to e.g. E64 (3684). The commissioning failed at BLE discovery since the manual paring code contains the E (E00 = 3584).

Parsed device manual pairing code:
```
chip_tool/chip-tool payload parse-setup-payload 34069117130
[1658752774.437685][17773:17773] CHIP:SPL: Parsing decimalRepresentation: 34069117130
[1658752774.437713][17773:17773] CHIP:SPL: Version:       0
[1658752774.437715][17773:17773] CHIP:SPL: VendorID:      0
[1658752774.437717][17773:17773] CHIP:SPL: ProductID:     0
[1658752774.437719][17773:17773] CHIP:SPL: Custom flow:   0    (STANDARD)
[1658752774.437720][17773:17773] CHIP:SPL: Capabilities:  0x00 (NONE)
[1658752774.437722][17773:17773] CHIP:SPL: Discriminator: 3584
[1658752774.437724][17773:17773] CHIP:SPL: Passcode:      28073715
```
Parsed device QR code payload:
```
chip-tool payload parse-setup-payload MT:KAYA3E.N16WWU-5KM00
[1658752813.867883][17774:17774] CHIP:SPL: Parsing base38Representation: MT:KAYA3E.N16WWU-5KM00
[1658752813.867921][17774:17774] CHIP:SPL: Version:       0
[1658752813.867924][17774:17774] CHIP:SPL: VendorID:      4735
[1658752813.867926][17774:17774] CHIP:SPL: ProductID:     32781
[1658752813.867928][17774:17774] CHIP:SPL: Custom flow:   0    (STANDARD)
[1658752813.867930][17774:17774] CHIP:SPL: Capabilities:  0x02 (BLE)
[1658752813.867933][17774:17774] CHIP:SPL: Discriminator: 3654
[1658752813.867936][17774:17774] CHIP:SPL: Passcode:      28073715
```

 #### Proposed Solution
The chip-tool during commissioning by manual pairing code shall connect the device based only on 4 most significant bits rather than the whole discriminator.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
A few changes here:

1. Implement a SetupDiscriminator class that commons up logic like "does this
   discriminator, which might be short or long, match the given long
   discriminator?" and "extract short discriminator from long discriminator".
2. Change SetupPayload to more clearly indicate whether it's storing a short or
   long discriminator, instead of storing a short discriminator the same way as
   a long discriminator with the low bits all 0 and hoping consumers check
   isShortDiscriminator.
3. Update BLE scanning APIs to take SetupDiscriminator.
4. Fix the Linux and Tizen BLE code to properly handle short discriminators
   (which used to not match if the long discriminator happened to have the low 8
   bits nonzero).
5. Fix the Darwin BLE code to properly handle long discriminators that have 0
   low bits.  Before this change they used to incorrectly match a long
   discriminator which had the same 4 high bits but different 8 low bits.

Fixes https://github.com/project-chip/connectedhomeip/issues/21160

#### Problem
See above.

#### Change overview
See above.

#### Testing
Ensured that BLE commissioning still works on Darwin (which is where I can test) and that chip-tool payload parsing commands work right.

### CODE REVIEW

File: src/lib/support/SetupDiscriminator.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        return mDiscriminator == LongToShortValue(discriminator);
+    }
+
+    bool operator==(const SetupDiscriminator & other) const
+    {
+        return mIsShortDiscriminator == other.mIsShortDiscriminator && mDiscriminator == other.mDiscriminator;
+    }
+
+private:
+    static_assert(kLongBits == 12, "Unexpected field length");
```
Comment (msandstedt) [inline src/lib/support/SetupDiscriminator.h:91]: Seems like we should have asserts for both kLongBits and kShortBits, or for neither.

File: src/lib/support/SetupDiscriminator.h [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        constexpr int kLongToShortShift = kLongBits - kShortBits;
+        return static_cast<uint8_t>(longValue >> kLongToShortShift);
+    }
+
+    // If long discriminator, all 12 bits are used.  If short discriminator,
+    // only the low kManualSetupDiscriminatorFieldLengthInBits bits are used, to
+    // store the value of the short discriminator (which contains only the high
+    // bits of the complete 12-bit discriminator).
+    uint16_t mDiscriminator : 12;
+    uint16_t mIsShortDiscriminator : 1;
```
Comment (msandstedt) [inline src/lib/support/SetupDiscriminator.h:108]:
```suggestion
    bool mIsShortDiscriminator;
```

File: src/lib/support/SetupDiscriminator.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/lib/support/SetupDiscriminator.h:91]: kLongBits needs to be asserted so we can fit in the 12-bit bitfield we are allocating for the discriminator.  The kShortBits value doesn't affect the validity of any of this code, though maybe we should assert it's small enough to fit in uint8_t and no larger than kLongBits.

**Review by Damian-Nordic — APPROVED** [review]

File: src/lib/support/SetupDiscriminator.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/lib/support/SetupDiscriminator.h]: Added assert for kShortBits too.

**Review by msandstedt — APPROVED** [review]

**Pull request merged**